src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -258,51 +258,91 @@ import scala.runtime.Statics
   *  @define evaluatesAllElements This method evaluates all elements of the collection.
   */
 @SerialVersionUID(3L)
-final class LazyList[+A] private(private[this] var lazyState: () => LazyList.State[A])
+final class LazyList[+A] private (lazyState: () => LazyList[A])
   extends AbstractSeq[A]
     with LinearSeq[A]
     with LinearSeqOps[A, LazyList, LazyList[A]]
     with IterableFactoryDefaults[A, LazyList]
     with Serializable {
   import LazyList._
 
-  @volatile private[this] var stateEvaluated: Boolean = false
-  @inline private def stateDefined: Boolean = stateEvaluated
-  private[this] var midEvaluation = false
-
-  private lazy val state: State[A] = {
-    // if it's already mid-evaluation, we're stuck in an infinite
-    // self-referential loop (also it's empty)
-    if (midEvaluation) {
-      throw new RuntimeException(
-        "LazyList evaluation depends on its own result (self-reference); see docs for more info"
-      )
+  private def this(head: A, tail: LazyList[A]) = {
+    this(LazyList.EmptyFun)
+    _head = head
+    _state = tail
+  }
+
+  // used to synchronize lazy state evaluation
+  // when `_head ne Uninitialized`
+  //   - `null` if this is an empty lazy list
+  //   - `head: A` otherwise (can be `null`, `_state == null` is used to test emptyness)
+  @volatile private[this] var _head: Any = Uninitialized
+
+  // when `_head eq Uninitialized`
+  //   - `lazySate: () => State`
+  //   - MidEvaluation while evaluating lazyState
+  // when `_head ne Uninitialized`
+  //   - `null` if this is an empty lazy list
+  //   - `tail: LazyList[A]` otherwise
+  private[this] var _state: AnyRef = lazyState
+
+  private def rawHead: Any = _head
+  private def rawState: AnyRef = _state
+
+  @inline private def stateDefined: Boolean = _head.asInstanceOf[AnyRef] ne Uninitialized
+
+  private def initState(): Unit = synchronized {
+    if (!stateDefined) {
+      // if it's already mid-evaluation, we're stuck in an infinite
+      // self-referential loop (also it's empty)
+      if (_state eq MidEvaluation)
+        throw new RuntimeException(
+          "LazyList evaluation depends on its own result (self-reference); see docs for more info")
+
+      val fun = _state.asInstanceOf[() => LazyList[A]]
+      _state = MidEvaluation
+      // restore `fun` in finally: consistent with previous 2.13 implementation which used a lazy val
+      val r = try fun.apply() finally _state = fun
+      r match {
+        case null =>
+          _state = null
+          _head = null
+        case l: LazyList[A @unchecked] =>
+          _state = l.rawState
+          _head = l.rawHead
+      }
```

Comment:
now that I'm looking at it, is this safe? should we make the assignments to `_state` and `_head` synchronized on `l` in case `l` is also being evaluated at the same time? ...is that secretly a disaster waiting to happen with an empty cyclical `LazyList`?

Reply:
to elaborate, you could interleave some series of events where `l.rawState` reads either `MidEvaluation` or `lazyState`, and then `l.rawHead` subsequently reads the evaluated head value for `l`, leaving `this._state` in an invalid/incorrect state

Reply:
You might be spot on. For example, if I look at `object cons`

```
def apply[A](hd: => A, tl: => LazyList[A]): LazyList[A] = newLL(eagerCons(hd, newLL(tl.state)))
```

we have `newLL(tl.state)`, which is `new LazyList[A](() => tl.state)`. Calling `state` method on `tl` guarantees initialization, but the types would just as well work out when writing `newLL(tl)`. Now there could be a race in reading `l.rawState` and `l.rawHead`.

Before we had the `State[A]` trait and its eager `Cons[A]` class.
  - there was a single field to assign
  - the `State` type ensured that `lazyState` would always return an eagerly evaluated `Cons`

Not sure if we can set up a test to reproduce the race.

Maybe the solution is to call `l.state` before assigning `_state` / `_head`? I'll keep looking.

Reply:
> Not sure if we can set up a test to reproduce the race.

it sounds annoying but it's probably doable if we manually create several locks and wait on them in very specific ways

> it prevents assigning non-evaluated fields to `_head` / `_state` in `initState`, which would be a bug

to clarify, would it be a bug because it would cause multiple evaluation?

Reply:
> > it prevents assigning non-evaluated fields to `_head` / `_state` in `initState`, which would be a bug
> 
> to clarify, would it be a bug because it would cause multiple evaluation?

You're right, `evaluated` calls itself after initialization so it would always return an initialized LazyList.